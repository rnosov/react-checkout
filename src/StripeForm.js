/*
 * StripeForm React Component
 *
 * Copyright Â© Roman Nosov 2016
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE.txt file in the root directory of this source tree.
 */
import React, { Component, PropTypes } from 'react';
import RestForm from 'redux-simpleform/dist/RestForm';

const
  propTypes = {
    endpoint: PropTypes.string.isRequired,
    testStripeKey: PropTypes.string.isRequired,
    liveStripeKey: PropTypes.string.isRequired,
    zeroAmountText: PropTypes.string,
    stripeNotLoadedText: PropTypes.string,
    waitText: PropTypes.string,
    errorText: PropTypes.string,
    successText: PropTypes.string,
    welcomeText: PropTypes.string,
    invalidFieldText: PropTypes.string,
    setStatus: PropTypes.func.isRequired,
  },
  defaultProps = {
    zeroAmountText: 'Your order is empty',
    stripeNotLoadedText: 'Unfortunately we are unable to process your payment',
    waitText: 'Processing your payment. Please wait ...',
    errorText: 'Houston, we have a problem!',
    successText: 'Your payment has been made',
    welcomeText: 'Please pay for your order',
    invalidFieldText: 'This field is invalid',
  };

class StripeForm extends Component {

  static isStripeLoading = false;

  static isStripe() {
    return typeof Stripe !== 'undefined';
  }

  componentDidMount() {
    if (StripeForm.isStripeLoading) return;
    StripeForm.isStripeLoading = true;
    setTimeout( () => {
      const script = document.createElement('script');
      script.src = 'https://js.stripe.com/v2/';
      script.async = true;
      document.body.appendChild(script);
    }, 500);
  }

  getStripeToken(card) {
    return new Promise((resolve, reject) => {
      Stripe.card.createToken(card, (status, { error, id }) => {
        if (error) {
          reject(error);
        } else {
          resolve(id);
        }
      });
    });
  }

  async handleFormWillFetch(form) {
    if (!StripeForm.isStripe())
      throw new Error( this.props.stripeNotLoadedText || 'Could not load Stripe.js' );
    const { number, cvc, exp, ...data } = form;
    const [ mode, stripeKey ] = process.env.NODE_ENV === 'production' ? ['live', this.props.liveStripeKey] : ['test', this.props.testStripeKey];
    Stripe.setPublishableKey(stripeKey);
    const token = await this.getStripeToken({ number, cvc, exp });
    return { mode, token, data: { ...data, ...this.props.order } };
  }

  async handleResponse(response) {
    const payload = await response.json();
    if (!payload.success)
      throw new Error(payload.errorMessage || payload.message || 'Unknown error');
    return this.props.successText;
  }

  handleParsingComplete(fields) {
    return fields.map( field => {
      switch (field.name) {
        case 'number': return field.merge({
          required: true,
          type: 'tel',
          autoComplete: 'cc-number',
          validate:  field => StripeForm.isStripe() ? this.validate(field, Stripe.card.validateCardNumber) : field,
          onChange: (newVal, oldVal) => {
            if (newVal && oldVal && newVal.length>oldVal.length) {
              if (/^\d{4}$/.test(newVal) || /^\d{4}\s\d{4}$/.test(newVal) || /^\d{4}\s\d{4}\s\d{4}$/.test(newVal))
                return newVal + ' ';
            }
            return /^[\d\s]*$/.test(newVal)?newVal:oldVal;
          },
        });
        case 'cvc': return field.merge({
          required: true,
          type: 'number',
          autoComplete: 'cc-csc',
          validate: field => StripeForm.isStripe() ? this.validate(field, Stripe.card.validateCVC) : field,
        });
        case 'exp': return field.merge({
          required: true,
          type: 'tel',
          autoComplete: 'cc-exp',
          validate: field => StripeForm.isStripe() ? this.validate(field, Stripe.card.validateExpiry) : field,
          onChange: (newVal, oldVal) => {
            if (newVal && oldVal && newVal.length>oldVal.length) {
              if (/^\d{2}$/.test(newVal))
                return newVal + '/';
              if (/^\d{2}\/\/$/.test(newVal))
                return oldVal;
            }
            return /^[\d\/]*$/.test(newVal)?newVal:oldVal;
          },
        });
      }
      return field;
    });
  }

  validate(field, func) {
    return func(field.value) ? field.set('error', false) : field
    .set('error', true)
    .set('message', this.props.invalidFieldText );
  }

  render() {
    const {
      testStripeKey,
      liveStripeKey,
      zeroAmountText,
      stripeNotLoadedText,
      amountPrefix,
      isDisplayed,
      invalidFieldText,
      order,
      ...props
    } = this.props;
    if (!isDisplayed)
      return <div className="alert alert-info"><p>{zeroAmountText}</p></div>;
    return(
      <RestForm
        {...props}
        onFormWillFetch={::this.handleFormWillFetch}
        onResponse={::this.handleResponse}
        onParsingComplete={::this.handleParsingComplete}
      />
    );
  }

}

StripeForm.propTypes = propTypes;
StripeForm.defaultProps = defaultProps;

export default StripeForm;
