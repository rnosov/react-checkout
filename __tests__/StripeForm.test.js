/*
 * StripeForm Test Suite
 *
 * Copyright Â© Roman Nosov 2016
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE.txt file in the root directory of this source tree.
 */

import React from 'react';
import { mount } from 'enzyme';
import StripeForm from '../src/StripeForm';

describe('StripeForm', () => {
  it('renders a initial view', () => {
    const form = mount(
      <StripeForm
        Name="*|John Doe"
        Email="*email|john@example.com"
        Phone="tel|+44 207 123 4567"
        number="*|4242 4242 4242 4242|Long number on the front of your card|Card Number"
        cvc="*|123|The 3 digits to the right of the signature strip located on the back of your card|CVC"
        exp="*|10/17||Expiry Date"
        Address="*6|1 Chapel Hill, Heswall, BOURNEMOUTH, UK, BH1 1AA|The address where your order will be shipped"
        setStatus={ () => void 0 }
        testStripeKey="123"
        liveStripeKey="123"
        endpoint="https://example.com/"
        order={{ description: 'donation' }}
        isDisplayed={true}
      />
    );
    expect(form.html()).toMatchSnapshot();
  });
});
