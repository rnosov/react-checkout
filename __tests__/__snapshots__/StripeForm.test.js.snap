exports[`StripeForm renders a initial view 1`] = `"<div class=\"simpleform-fadeIn\"><p>Please pay for your order</p><form novalidate=\"\" class=\"m-x-1 text-xs-left\" style=\"overflow: hidden;\"><noscript>Please enable javascript in order to use this form.</noscript><div class=\"form-group row\"><label class=\"col-xs-12 text-sm-right col-form-label col-sm-2\">*Name</label><div class=\"col-xs-12 col-sm-5\"><input type=\"text\" class=\"form-control\" value=\"\" data-id=\"0\" placeholder=\"John Doe\"></div><!-- react-empty: 9 --></div><div class=\"form-group row\"><label class=\"col-xs-12 text-sm-right col-form-label col-sm-2\">*Email</label><div class=\"col-xs-12 col-sm-5\"><input type=\"email\" class=\"form-control\" value=\"\" data-id=\"1\" placeholder=\"john@example.com\"></div><!-- react-empty: 14 --></div><div class=\"form-group row\"><label class=\"col-xs-12 text-sm-right col-form-label col-sm-2\">Phone</label><div class=\"col-xs-12 col-sm-5\"><input type=\"tel\" class=\"form-control\" value=\"\" data-id=\"2\" placeholder=\"+44 207 123 4567\"></div><!-- react-empty: 19 --></div><div class=\"form-group row\"><label class=\"col-xs-12 text-sm-right col-form-label col-sm-2\">*Card Number</label><div class=\"col-xs-12 col-sm-5\"><input type=\"tel\" class=\"form-control\" value=\"\" data-id=\"3\" autocomplete=\"cc-number\" placeholder=\"4242 4242 4242 4242\"><small class=\"form-text text-muted\">Long number on the front of your card</small></div><!-- react-empty: 25 --></div><div class=\"form-group row\"><label class=\"col-xs-12 text-sm-right col-form-label col-sm-2\">*CVC</label><div class=\"col-xs-12 col-sm-5\"><input type=\"number\" class=\"form-control\" value=\"\" data-id=\"4\" autocomplete=\"cc-csc\" placeholder=\"123\"><small class=\"form-text text-muted\">The 3 digits to the right of the signature strip located on the back of your card</small></div><!-- react-empty: 31 --></div><div class=\"form-group row\"><label class=\"col-xs-12 text-sm-right col-form-label col-sm-2\">*Expiry Date</label><div class=\"col-xs-12 col-sm-5\"><input type=\"tel\" class=\"form-control\" value=\"\" data-id=\"5\" autocomplete=\"cc-exp\" placeholder=\"10/17\"></div><!-- react-empty: 36 --></div><div class=\"form-group row\"><label class=\"col-xs-12 text-sm-right col-form-label col-sm-2\">*Address</label><div class=\"col-xs-12 col-sm-5\"><textarea class=\"form-control\" data-id=\"6\" rows=\"6\" placeholder=\"1 Chapel Hill, Heswall, BOURNEMOUTH, UK, BH1 1AA\"></textarea><small class=\"form-text text-muted\">The address where your order will be shipped</small></div><!-- react-empty: 42 --></div><div class=\"form-group row\"><div class=\"col-xs-12 col-sm-2\"><small class=\"form-text text-muted text-sm-right form-control-static\"><!-- react-text: 46 -->*<!-- /react-text --><!-- react-text: 47 -->Mandatory fields<!-- /react-text --></small></div><div class=\"col-xs-12 form-inline col-sm-8\"><div class=\"form-group has-danger\"><button type=\"submit\" class=\"btn btn-outline-primary\">Submit</button><div class=\"text-muted form-control-static\">&nbsp;&nbsp; </div><!-- react-empty: 52 --></div></div></div></form></div>"`;
